### DISCUSSION TITLE
fails at linking step when compiling lammps

### DISCUSSION THREAD

**Discussion by zfan2016** [deepmodeling/deepmd-kit#222] [Q&A]
We built both python and c-- interfaces for both tensorflow and deepmd successfully.
When compiled lammps with deepmd packages, it failed at linking step:

Our compiling step and error message are below:

1006 git clone https://github.com/lammps/lammps.git
1007 cd lammps
1008 cd src/
1009 cp -r ../../deepmd-kit/source/build/USER-DEEPMD/ .
1010 make yes-kspace
1011 make make yes-user-deepmd
1012 make yes-user-deepmd
1013 make -j8 mpi
1014 module load ompi
1015 module load mvapich
1016 vim Makefile
1017 vim MAKE/Makefile.mpi
1018 make clean
1019 make -j8 mpi

CC -g -O3 -DLAMMPS_GZIP -DLAMMPS_MEMALIGN=64 -std=c++11 -DHIGH_PREC -I/global/cscratch1/sd/rgayatri/ConsultingTickets/INC0150840/TF/tensorflow/tensorflow_root/include -I/global/cscratch1/sd/rgayatri/ConsultingTickets/INC0150840/TF/tensorflow/tensorflow_root/include -I/global/cscratch1/sd/rgayatri/ConsultingTickets/INC0150840/deepmd-kit/source/build/include/deepmd -DMPICH_SKIP_MPICXX -DOMPI_SKIP_MPICXX=1 -c ../angle.cpp
ar -rc ../liblammps_mpi.a write_restart.o write_dump.o write_data.o write_coeff.o verlet.o velocity.o variable.o utils.o update.o universe.o timer.o thermo.o special.o set.o run.o respa.o reset_ids.o rerun.o replicate.o remap_wrap.o remap.o region_union.o region_sphere.o region_prism.o region_plane.o region_intersect.o region_deprecated.o region_cylinder.o region_cone.o region_block.o region.o reader_xyz.o reader_native.o reader.o read_restart.o read_dump.o read_data.o rcb.o random_park.o random_mars.o procmap.o pppm_tip4p.o pppm_stagger.o pppm_dplr.o pppm_disp_tip4p.o pppm_disp.o pppm_dipole_spin.o pppm_dipole.o pppm_cg.o pppm.o pair_zero.o pair_zbl.o pair_yukawa.o pair_ufm.o pair_tip4p_long.o pair_table.o pair_soft.o pair_nnp.o pair_morse.o pair_mie_cut.o pair_lj_smooth_linear.o pair_lj_smooth.o pair_lj_long_tip4p_long.o pair_lj_long_coul_long.o pair_lj_gromacs_coul_gromacs.o pair_lj_gromacs.o pair_lj_expand.o pair_lj_cut_tip4p_long.o pair_lj_cut_coul_wolf.o pair_lj_cut_coul_msm.o pair_lj_cut_coul_long.o pair_lj_cut_coul_dsf.o pair_lj_cut_coul_debye.o pair_lj_cut_coul_cut.o pair_lj_cut.o pair_lj_cubic.o pair_lj_charmmfsw_coul_long.o pair_lj_charmm_coul_msm.o pair_lj_charmm_coul_long.o pair_lj96_cut.o pair_hybrid_overlay.o pair_hybrid.o pair_gauss.o pair_dpd_tstat.o pair_dpd.o pair_deprecated.o pair_coul_wolf.o pair_coul_streitz.o pair_coul_msm.o pair_coul_long.o pair_coul_dsf.o pair_coul_debye.o pair_coul_cut.o pair_buck_long_coul_long.o pair_buck_coul_msm.o pair_buck_coul_long.o pair_buck_coul_cut.o pair_buck.o pair_born_coul_wolf.o pair_born_coul_msm.o pair_born_coul_long.o pair_born_coul_dsf.o pair_born.o pair_beck.o pair.o output.o ntopo_improper_template.o ntopo_improper_partial.o ntopo_improper_all.o ntopo_dihedral_template.o ntopo_dihedral_partial.o ntopo_dihedral_all.o ntopo_bond_template.o ntopo_bond_partial.o ntopo_bond_all.o ntopo_angle_template.o ntopo_angle_partial.o ntopo_angle_all.o ntopo.o nstencil_half_multi_3d_newton_tri.o nstencil_half_multi_3d_newton.o nstencil_half_multi_3d_newtoff.o nstencil_half_multi_2d_newton_tri.o nstencil_half_multi_2d_newton.o nstencil_half_multi_2d_newtoff.o nstencil_half_ghost_bin_3d_newtoff.o nstencil_half_ghost_bin_2d_newtoff.o nstencil_half_bin_3d_newton_tri.o nstencil_half_bin_3d_newton.o nstencil_half_bin_3d_newtoff.o nstencil_half_bin_2d_newton_tri.o nstencil_half_bin_2d_newton.o nstencil_half_bin_2d_newtoff.o nstencil_full_multi_3d.o nstencil_full_multi_2d.o nstencil_full_ghost_bin_3d.o nstencil_full_ghost_bin_2d.o nstencil_full_bin_3d.o nstencil_full_bin_2d.o nstencil.o npair_skip_size_off2on_oneside.o npair_skip_size_off2on.o npair_skip_size.o npair_skip_respa.o npair_skip.o npair_halffull_newton.o npair_halffull_newtoff.o npair_half_size_nsq_newton.o npair_half_size_nsq_newtoff.o npair_half_size_bin_newton_tri.o npair_half_size_bin_newton.o npair_half_size_bin_newtoff.o npair_half_respa_nsq_newton.o npair_half_respa_nsq_newtoff.o npair_half_respa_bin_newton_tri.o npair_half_respa_bin_newton.o npair_half_respa_bin_newtoff.o npair_half_nsq_newton.o npair_half_nsq_newtoff_ghost.o npair_half_nsq_newtoff.o npair_half_multi_newton_tri.o npair_half_multi_newton.o npair_half_multi_newtoff.o npair_half_bin_newton_tri.o npair_half_bin_newton.o npair_half_bin_newtoff_ghost.o npair_half_bin_newtoff.o npair_half_bin_atomonly_newton.o npair_full_nsq_ghost.o npair_full_nsq.o npair_full_multi.o npair_full_bin_ghost.o npair_full_bin_atomonly.o npair_full_bin.o npair_copy.o npair.o neighbor.o neigh_request.o neigh_list.o nbin_standard.o nbin.o msm_cg.o msm.o molecule.o modify.o minimize.o min_sd.o min_quickmin.o min_linesearch.o min_hftn.o min_fire_old.o min_fire.o min_cg.o min.o memory.o math_special.o math_extra.o lmppython.o library.o lattice.o lammps.o kspace_deprecated.o kspace.o irregular.o integrate.o input.o info.o improper_zero.o improper_hybrid.o improper_deprecated.o improper.o imbalance_var.o imbalance_time.o imbalance_store.o imbalance_neigh.o imbalance_group.o imbalance.o image.o hashlittle.o group.o gridcomm.o force.o fix_wall_region.o fix_wall_reflect.o fix_wall_morse.o fix_wall_lj93.o fix_wall_lj126.o fix_wall_lj1043.o fix_wall_harmonic.o fix_wall.o fix_viscous.o fix_vector.o fix_tune_kspace.o fix_tmd.o fix_temp_rescale.o fix_temp_csvr.o fix_temp_csld.o fix_temp_berendsen.o fix_store_state.o fix_store_force.o fix_store.o fix_spring_self.o fix_spring_rg.o fix_spring_chunk.o fix_spring.o fix_setforce.o fix_restrain.o fix_respa.o fix_recenter.o fix_read_restart.o fix_property_atom.o fix_print.o fix_press_berendsen.o fix_planeforce.o fix_nvt_sphere.o fix_nvt_sllod.o fix_nvt.o fix_nve_sphere.o fix_nve_noforce.o fix_nve_limit.o fix_nve.o fix_numdiff.o fix_npt_sphere.o fix_npt.o fix_nph_sphere.o fix_nph.o fix_nh_sphere.o fix_nh.o fix_neigh_history.o fix_move.o fix_momentum.o fix_minimize.o fix_lineforce.o fix_langevin.o fix_indent.o fix_heat.o fix_halt.o fix_group.o fix_gravity.o fix_external.o fix_enforce2d.o fix_dummy.o fix_dt_reset.o fix_drag.o fix_dplr.o fix_deprecated.o fix_deform.o fix_controller.o fix_box_relax.o fix_balance.o fix_aveforce.o fix_ave_time.o fix_ave_histo_weight.o fix_ave_histo.o fix_ave_correlate.o fix_ave_chunk.o fix_ave_atom.o fix_addforce.o fix_adapt.o fix.o finish.o fft3d_wrap.o fft3d.o ewald_disp.o ewald_dipole_spin.o ewald_dipole.o ewald.o error.o dump_xyz.o dump_movie.o dump_local.o dump_image.o dump_deprecated.o dump_dcd.o dump_custom.o dump_cfg.o dump_atom.o dump.o domain.o displace_atoms.o dihedral_zero.o dihedral_hybrid.o dihedral_deprecated.o dihedral.o deprecated.o delete_bonds.o delete_atoms.o create_box.o create_bonds.o create_atoms.o compute_vcm_chunk.o compute_vacf.o compute_torque_chunk.o compute_temp_sphere.o compute_temp_region.o compute_temp_ramp.o compute_temp_profile.o compute_temp_partial.o compute_temp_deform.o compute_temp_com.o compute_temp_chunk.o compute_temp.o compute_stress_atom.o compute_slice.o compute_reduce_region.o compute_reduce_chunk.o compute_reduce.o compute_rdf.o compute_property_local.o compute_property_chunk.o compute_property_atom.o compute_pressure.o compute_pe_atom.o compute_pe.o compute_pair_local.o compute_pair.o compute_orientorder_atom.o compute_omega_chunk.o compute_msd_chunk.o compute_msd.o compute_ke_atom.o compute_ke.o compute_inertia_chunk.o compute_improper_local.o compute_improper.o compute_hexorder_atom.o compute_heat_flux.o compute_gyration_chunk.o compute_gyration.o compute_group_group.o compute_global_atom.o compute_fragment_atom.o compute_erotate_sphere_atom.o compute_erotate_sphere.o compute_displace_atom.o compute_dipole_chunk.o compute_dihedral_local.o compute_dihedral.o compute_deprecated.o compute_coord_atom.o compute_contact_atom.o compute_com_chunk.o compute_com.o compute_cna_atom.o compute_cluster_atom.o compute_chunk_spread_atom.o compute_chunk_atom.o compute_centroid_stress_atom.o compute_centro_atom.o compute_bond_local.o compute_bond.o compute_angmom_chunk.o compute_angle_local.o compute_angle.o compute_aggregate_atom.o compute_adf.o compute.o comm_tiled.o comm_brick.o comm.o citeme.o change_box.o bond_zero.o bond_hybrid.o bond_deprecated.o bond.o body.o balance.o atom_vec_tri.o atom_vec_sphere.o atom_vec_line.o atom_vec_hybrid.o atom_vec_ellipsoid.o atom_vec_charge.o atom_vec_body.o atom_vec_atomic.o atom_vec.o atom_map.o atom.o angle_zero.o angle_hybrid.o angle_deprecated.o angle.o
CC -g -O3 main.o -L/global/cscratch1/sd/rgayatri/ConsultingTickets/INC0150840/TF/tensorflow/tensorflow_root/lib -L/global/cscratch1/sd/rgayatri/ConsultingTickets/INC0150840/TF/tensorflow/tensorflow_root/lib -L/global/cscratch1/sd/rgayatri/ConsultingTickets/INC0150840/deepmd-kit/source/build/lib -L. -llammps_mpi -Wl,--no-as-needed -ldeepmd_op -ldeepmd -ltensorflow_cc -ltensorflow_framework -Wl,-rpath=/global/cscratch1/sd/rgayatri/ConsultingTickets/INC0150840/TF/tensorflow/tensorflow_root/lib -Wl,-rpath=/global/cscratch1/sd/rgayatri/ConsultingTickets/INC0150840/TF/tensorflow/tensorflow_root/lib -Wl,-rpath=/global/cscratch1/sd/rgayatri/ConsultingTickets/INC0150840/deepmd-kit/source/build/lib -o ../lmp_mpi
/usr/bin/ld: cannot find -ltensorflow_framework
make[1]: *** [Makefile:89: ../lmp_mpi] Error 1
make[1]: Leaving directory '/global/cscratch1/sd/lammps/src/Obj_mpi'
make: *** [Makefile:207: mpi] Error 2

Can you give us some suggestions to resolve this issue? Thanks.

**Comment by njzjz** [comment]
Can you ensure you have something like `libtensorflow_framework.so` in /global/cscratch1/sd/rgayatri/ConsultingTickets/INC0150840/TF/tensorflow/tensorflow_root/lib?

**Comment by njzjz** [comment]
What is your Makefile.package like?

**Comment by zfan2016** [comment]
In the lammps/src/MAKEMakefil.mpi, I replace mpicxx with CC as CC and LINK variables.

**Comment by rgayatri23** [comment]
Hi, Yes I have the libtensorflow_framework.so in both lib and include directories but still get the same error. 
rgayatri@cori09:/global/cscratch1/sd/rgayatri/ConsultingTickets/INC0150840/lammps/src> !make
make -j8 mpi
Gathering installed package information (may take a little while)
make[1]: warning: -jN forced in submake: disabling jobserver mode.
make[1]: Entering directory '/global/cscratch1/sd/rgayatri/ConsultingTickets/INC0150840/lammps/src'
make[1]: 'lmpinstalledpkgs.h' is up to date.
Gathering git version information
make[1]: Leaving directory '/global/cscratch1/sd/rgayatri/ConsultingTickets/INC0150840/lammps/src'
Compiling LAMMPS for machine mpi
make[1]: warning: -jN forced in submake: disabling jobserver mode.
make[1]: Entering directory '/global/cscratch1/sd/rgayatri/ConsultingTickets/INC0150840/lammps/src/Obj_mpi'
make[1]: Leaving directory '/global/cscratch1/sd/rgayatri/ConsultingTickets/INC0150840/lammps/src/Obj_mpi'
make[1]: warning: -jN forced in submake: disabling jobserver mode.
make[1]: Entering directory '/global/cscratch1/sd/rgayatri/ConsultingTickets/INC0150840/lammps/src/Obj_mpi'
CC -g -O3 main.o -L/global/cscratch1/sd/rgayatri/ConsultingTickets/INC0150840/TF/tensorflow/tensorflow_root/lib -L/global/cscratch1/sd/rgayatri/ConsultingTickets/INC0150840/TF/tensorflow/tensorflow_root/lib -L/global/cscratch1/sd/rgayatri/ConsultingTickets/INC0150840/deepmd-kit/source/build/lib     -L. -llammps_mpi -Wl,--no-as-needed -ldeepmd_op -ldeepmd -ltensorflow_cc -ltensorflow_framework -Wl,-rpath=/global/cscratch1/sd/rgayatri/ConsultingTickets/INC0150840/TF/tensorflow/tensorflow_root/lib -Wl,-rpath=/global/cscratch1/sd/rgayatri/ConsultingTickets/INC0150840/TF/tensorflow/tensorflow_root/lib -Wl,-rpath=/global/cscratch1/sd/rgayatri/ConsultingTickets/INC0150840/deepmd-kit/source/build/lib      -o ../lmp_mpi
/usr/bin/ld: cannot find -ltensorflow_framework
make[1]: *** [Makefile:89: ../lmp_mpi] Error 1
make[1]: Leaving directory '/global/cscratch1/sd/rgayatri/ConsultingTickets/INC0150840/lammps/src/Obj_mpi'
make: *** [Makefile:207: mpi] Error 2
rgayatri@cori09:/global/cscratch1/sd/rgayatri/ConsultingTickets/INC0150840/lammps/src> ls ../../TF/tensorflow/tensorflow_root/include/
Eigen  absl  external  google  libtensorflow_framework.so  solaris  tensorflow  third_party  unsupported
rgayatri@cori09:/global/cscratch1/sd/rgayatri/ConsultingTickets/INC0150840/lammps/src> ls ../../TF/tensorflow/tensorflow_root/lib/
libtensorflow_cc.so    libtensorflow_cc.so.1.14.0           libtensorflow_framework.so    libtensorflow_framework.so.1.14.0           src
libtensorflow_cc.so.1  libtensorflow_cc.so.1.14.0-2.params  libtensorflow_framework.so.1  libtensorflow_framework.so.1.14.0-2.params  tensorflow_framework
rgayatri@cori09:/global/cscratch1/sd/rgayatri/ConsultingTickets/INC0150840/lammps/src>

**Comment by joshgabriel** [comment]
Hi,

I am encountering the same issue. The Makefile.Package is like below and please note that $tensorflow_root and $deepmd_root are fully expanded paths in the actual file:

![image](https://user-images.githubusercontent.com/13017592/81369114-3886e180-90b7-11ea-832f-29a679122e22.png)

Thanks for any suggestions,
Joshua

**Comment by joshgabriel** [comment]
I found a possible resolution for this issue, at least get a successful build and lammps script runs with a deepmd potential. 
1. Build with Lammps Aug 2019
2. Link path/to/tensorflow_install/lib/libtensorflow_cc.so to path/to/tensorflow_install/lib/libtensorflow_framework.so.1
with ln -s
